Type hint for \DateTimeInterface instead of \DateTime

jameshalsall opened the pull request:
This change allows userland code to pass instances of \DateTimeImmutable as well as \DateTime by using \DateTimeInterface in type hints instead of the more strict \DateTime.

Commit:
Type hint for \DateTimeInterface instead of \DateTime

This change allows userland code to pass instances of \DateTimeImmutable as
well as \DateTime by using \DateTimeInterface in type hints instead of the
more strict \DateTime.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=7174) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=7174) before we can accept your contribution.<br/>

schmengler commented:
Thank you for your contribution! Did you try to actually exchange `DateTime` with `DateTimeImmutable` after the change?  I'm still reviewing possible problems (Magento might use methods that are only present in `DateTime`)

adragus-inviqa commented:
> Magento might use methods that are only present in DateTime

@schmengler - there are no methods on `\DateTime` which aren't available in `\DateTimeImmutable`. It's the other way around, actually (the `DateTimeImmutable::createFromMutable()` method). 

Though there might be subtle changes/bugs in M2's behaviour, that would just point out the fact that mutable states create problems.

orlangur commented:
Nice! :) Was thinking of such change and wasn't aware there is even PR proposed already.

The only thing missing as to me is static test enforcement, so that `\DateTime` never ever used in type-hinting and PHPDoc. We already have some legacy tests checking obsolete ZF classes are not used/used only in particular place.

schmengler commented on lib/internal/Magento/Framework/Stdlib/DateTime/Timezone.php:
```diff
@@ -162,7 +162,7 @@ public function date($date = null, $locale = null, $useTimezone = true)
             return new \DateTime('now', new \DateTimeZone($timezone));
-        } elseif ($date instanceof \DateTime) {
+        } elseif ($date instanceof \DateTimeInterface) {
             return $date->setTimezone(new \DateTimeZone($timezone));
```
`setTimezone` is not part of `DateTimeInterface`

jameshalsall replied:
I think the only thing we can do here for best DX is to internally convert the `DateTimeImmutable` to `DateTime` and enforce the return type hint of `DateTime` always

schmengler commented:
> there are no methods on \DateTime which aren't available in \DateTimeImmutable. It's the other way around, actually (the DateTimeImmutable::createFromMutable() method).

Yes, but there are two problems:

1. The "setters" in `DateTimeImmutable` work different than those in `DateTime`
2. For this reason, they are not part of `DateTimeInterface`

Type hinting against `DateTimeInterface` is perfect where only the interface methods are used (see http://php.net/manual/en/class.datetimeinterface.php), like formatting methods.

But you cannot replace `DateTime` with `DateTimeInterface` where methods like `setTimestamp()` are used. It's possible to make these methods compatible with `DateTime` and `DateTimeImmutable` with code like

```
    $date = $date->setTimezone($tz);
```

but then check explicitly for `DateTime` and `DateTimeImmutable`, otherwise the code will break with userland implementations and test dummies.

Can you please review your changes again? I found at least one occurence of  `setTimezone` after a check for `DateTimeInterface` (see inline comment)

jameshalsall commented:
I'll take another look at this and revert the typehints where it would break code

jameshalsall commented:
Haven't forgotten about this :)

Vinai commented:
Please also add a `<preference for="DateTimeInterface" type="DateTime"/>`. I think `app/etc/di.xml` would be the right place.  
Using `DateTime` instead of `DateTimeImmutable` as the default preference would be better for backward compatibility.  
Also, I'm missing tests. Maybe https://github.com/magento/magento2/pull/4489 can serve as an example.

Commit:
Add unit tests for Timezone in stdlib, fix edge cases

Commit:
Add di.xml preference for DateTimeInterface

Commit:
Update DateTime methods in stdlib to support DateTimeInterface

Commit:
Fix some incorrect typehints in report collection

Commit:
Fix broken typehint and test

Commit:
Ensure Timezone set correctly in Timezone test

Commit:
Test tidy up

Commit:
Make expected result a callable in Timezone dataprovider

schmengler commented:
@jameshalsall Thanks for the updates! Just wanted to check back with you, if you plan to work on the failing tests or need any help?

okorshenko commented:
Hi @jameshalsall 
Could you please resolve merge conflicts so that we can proceed with code review?

jameshalsall commented:
This has been updated now. 

@schmengler - the test failures are unrelated.

ishakhsuvarov committed:
magento/magento2#7174: Merge branch 'develop' of github.com:magento/magento2 into use-date-time-interface-in-instanceof-checks

ishakhsuvarov commented:
@jameshalsall I believe test failures are related, since the tests in question are green on the `develop` branch. I've also ran builds in the internal CICD and see the same result. 
However, I could not yet reproduce these failures on my local environment, may be the issue is somehow related to the timezone settings, not sure yet. Let us know if you have any ideas

ishakhsuvarov commented:
Actually, test failures identical to the build are reproducible if you run the whole suite of unit tests, for instance `bin/magento dev:tests:run unit`. I will continue investigating the cause of this problem.

ishakhsuvarov committed:
magento/magento2#7174: Merge branch 'develop' of github.com:magento/magento2 into use-date-time-interface-in-instanceof-checks

schmengler commented:
Thanks for chiming in @ishakhsuvarov

ishakhsuvarov committed:
magento/magento2#7174: Type hint for \DateTimeInterface instead of \DateTime

 - Updated unit test which changed default timezone and affected all test suite
 - Updated copyright to the up-to-date version

ishakhsuvarov commented:
@jameshalsall @schmengler 
Looks like the issue which affected multiple tests in the suite was located in the `\Magento\Framework\Stdlib\Test\Unit\DateTime\TimezoneTest` as it was calling the `date_default_timezone_set` inside the dataProvider, which led to the incorrect rollback of the timezone setting after test execution. 

Since DataProvider seems to be executed before setUp I had to slightly shuffle the order of things in the unit test. Currently it does not look as good and logical as it should be, however I've pushed it to make sure builds are passing. We can improve it afterwards.

ishakhsuvarov committed:
magento/magento2#7174: Type hint for \DateTimeInterface instead of \DateTime

     - Updated copyright to the up-to-date version

jameshalsall commented:
@ishakhsuvarov thanks for fixing, I'm pretty sure I had addressed that previously but I may have removed it by accident

ishakhsuvarov committed:
magento/magento2#7174: Merge branch 'develop' of github.com:magento/magento2 into use-date-time-interface-in-instanceof-checks
